gazebo_plugins/src/gazebo_ros_camera.cpp

```diff
@@ -90,34 +139,88 @@ GazeboRosCamera::~GazeboRosCamera()
 {
-  gazebo::CameraPlugin::Load(_sensor, _sdf);
+  // Initialize ROS node
+  impl_->ros_node_ = gazebo_ros::Node::Get(_sdf);
+
+  if (std::dynamic_pointer_cast<gazebo::sensors::DepthCameraSensor>(_sensor)) {
+    impl_->sensor_type_ = GazeboRosCameraPrivate::DEPTH;
+    gazebo::DepthCameraPlugin::Load(_sensor, _sdf);
+  } else if (std::dynamic_pointer_cast<gazebo::sensors::CameraSensor>(_sensor)) {
+    impl_->sensor_type_ = GazeboRosCameraPrivate::CAMERA;
+    gazebo::CameraPlugin::Load(_sensor, _sdf);
+  } else {
+    RCLCPP_ERROR(impl_->ros_node_->get_logger(),
+      "Plugin must be attached to sensor of type `camera` or `depth`");
+    impl_->ros_node_.reset();
+    return;
+  }
 
   // Camera name
   impl_->camera_name_ = _sdf->Get<std::string>("camera_name", _sensor->Name()).first;
 
   // Get tf frame for output
   impl_->frame_name_ = gazebo_ros::SensorFrameID(*_sensor, *_sdf);
 
-  // Initialize ROS node
-  impl_->ros_node_ = gazebo_ros::Node::Get(_sdf);
+  if (impl_->sensor_type_ == GazeboRosCameraPrivate::CAMERA) {
+    // Image publisher
+    // TODO(louise) Migrate image_connect logic once SubscriberStatusCallback is ported to ROS2
+    impl_->image_pub_ = image_transport::create_publisher(impl_->ros_node_.get(),
+        impl_->camera_name_ + "/image_raw");
 
-  // Image publisher
-  // TODO(louise) Migrate image_connect logic once SubscriberStatusCallback is ported to ROS2
-  impl_->image_pub_ = image_transport::create_publisher(impl_->ros_node_.get(),
-      impl_->camera_name_ + "/image_raw");
+    // TODO(louise) Uncomment this once image_transport::Publisher has a function to return the
+    // full topic.
+    // RCLCPP_INFO(impl_->ros_node_->get_logger(), "Publishing images to [%s]",
+    //  impl_->image_pub_.getTopic());
 
-  // TODO(louise) Uncomment this once image_transport::Publisher has a function to return the
-  // full topic.
-  // RCLCPP_INFO(impl_->ros_node_->get_logger(), "Publishing images to [%s]",
-  //  impl_->image_pub_.getTopic());
+    // Camera info publisher
+    // TODO(louise) Migrate ImageConnect logic once SubscriberStatusCallback is ported to ROS2
+    impl_->camera_info_pub_ = impl_->ros_node_->create_publisher<sensor_msgs::msg::CameraInfo>(
+      impl_->camera_name_ + "/camera_info", rclcpp::SensorDataQoS());
 
-  // Camera info publisher
-  // TODO(louise) Migrate ImageConnect logic once SubscriberStatusCallback is ported to ROS2
-  impl_->camera_info_pub_ = impl_->ros_node_->create_publisher<sensor_msgs::msg::CameraInfo>(
-    impl_->camera_name_ + "/camera_info", rclcpp::SensorDataQoS());
+    RCLCPP_INFO(impl_->ros_node_->get_logger(), "Publishing camera info to [%s]",
+      impl_->camera_info_pub_->get_topic_name());
+
+  } else {
+    // Image publisher
+    // TODO(shivesh) Migrate image_connect logic once SubscriberStatusCallback is ported to ROS2
+    impl_->image_pub_ = image_transport::create_publisher(impl_->ros_node_.get(),
+        impl_->camera_name_ + "/ir/image_raw");
```

Comment:
On the Kinect sensor, images prefixed with `ir` correspond to the Infrared sensor readings, which [look different](http://wiki.ros.org/openni_launch/Tutorials/IntrinsicCalibration#IR_.28depth.29_camera) from the raw camera image.

I saw that the `gazebo_ros_openni_kinect` plugin had topics with `ir`, but I think they were never used. So I think we can just skip publishing the raw images for depth cameras.

Reply:
Raw rgb images from depth cameras did have use cases in ROS1. I don't think we should remove it. Also the pointcloud's rgb colour is assigned from these images. So isn't it almost similar to RGB images?

Reply:
> Raw rgb images from depth cameras did have use cases in ROS1

Wait, were depth cameras publishing RGB images too?

> So isn't it almost similar to RGB images?

Yup, I agree it's almost the same. And the image is available, so we might as well publish it. I just think we don't *need* to, because afaik ROS 1 didn't publish them. But it doesn't hurt to do so. We just need a better topic name than `ir`. Maybe it could publish in the same topic as an RGB camera would (i.e. `<camera name>/image_raw`). What do you think?

Reply:
Depth camera did publish RGB images https://github.com/ros-simulation/gazebo_ros_pkgs/blob/3cd100e20ba09245be01ad674248aa31a46ad271/gazebo_plugins/src/gazebo_ros_depth_camera.cpp#L83
Even I would prefer changing the topic name

Reply:
I saw that topic being set, but it doesn't look like it's used anywhere, right?

Reply:
https://github.com/ros-simulation/gazebo_ros_pkgs/blob/3cd100e20ba09245be01ad674248aa31a46ad271/gazebo_plugins/src/gazebo_ros_depth_camera.cpp#L312
This calls `PutCameraData` in [GazeboRosCameraUtils](https://github.com/ros-simulation/gazebo_ros_pkgs/blob/kinetic-devel/gazebo_plugins/src/gazebo_ros_camera_utils.cpp) which publishes the image

Reply:
Oh you're totally right! Nevermind, sorry for the noise.

> Even I would prefer changing the topic name

+1, let's do that and point it out on the migration notes